x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/merge_unwired_stream_pipeline_actions.ts

```diff
@@ -0,0 +1,338 @@
+        ...pipeline,
+        processors,
+      },
+    });
+  } else {
+    actionsByType.delete_ingest_pipeline.push({
+      type: 'delete_ingest_pipeline',
+      request: {
+        name: pipelineName,
+      },
+    });
+
+    const targetTemplateNames = uniq(actions.map((action) => action.template));
+    if (targetTemplateNames.length !== 1) {
+      throw new Error(
+        'Actions targeting the same existing Streams managed pipeline target different templates'
+      );
+    }
+    const indexTemplate = await getIndexTemplate(targetTemplateNames[0], scopedClusterClient);
+
+    actionsByType.upsert_index_template.push({
+      type: 'upsert_index_template',
+      request: {
+        name: indexTemplate.name,
+        ...indexTemplate.index_template,
+        ignore_missing_component_templates:
+          castArray(indexTemplate.index_template.ignore_missing_component_templates) ?? [],
+        template: {
+          ...(indexTemplate.index_template.template ?? {}),
+          settings: {
+            ...(indexTemplate.index_template.template?.settings ?? {}),
+            default_pipeline: undefined,
+          },
+        },
+      },
+    });
+
+    actionsByType.upsert_write_index_or_rollover.push(
+      ...actions.map<UpsertWriteIndexOrRolloverAction>((action) => ({
+        type: 'upsert_write_index_or_rollover',
+        request: {
+          name: action.dataStream,
+        },
+      }))
+    );
+  }
+}
+
+async function updateExistingUserManagedPipeline({
+  pipelineName,
+  actions,
+  actionsByType,
+  scopedClusterClient,
+}: {
+  pipelineName: string;
+  actions: UnwiredStreamPipelineAction[];
+  actionsByType: ActionsByType;
+  scopedClusterClient: IScopedClusterClient;
+}) {
+  const referencePipelineNames = actions.map((action) => action.referencePipeline);
+  const { targetPipelineName, targetPipeline } = await findPipelineToModify(
+    pipelineName,
+    referencePipelineNames,
+    scopedClusterClient
+  );
+
+  let processors = targetPipeline.processors ?? [];
+  const existingPipelineProcessors = processors
+    .filter((processor) => processor.pipeline !== undefined)
+    .map((processor) => processor.pipeline?.name);
+
+  for (const action of actions) {
+    if (action.type === 'append_processor_to_ingest_pipeline') {
+      if (!existingPipelineProcessors.includes(action.processor.pipeline?.name)) {
+        processors.push(action.processor);
+      }
+    } else {
+      processors = processors.filter(
+        (processor) => processor.pipeline && processor.pipeline.name === action.referencePipeline
+      );
+    }
+  }
+
+  actionsByType.upsert_ingest_pipeline.push({
+    type: 'upsert_ingest_pipeline',
+    // All of these are UnwiredStreams so take any stream name to use for the ordering of operations
+    stream: actions[0].dataStream,
+    request: {
+      id: targetPipelineName,
+      ...targetPipeline,
+      processors,
+    },
+  });
+}
+
+async function findPipelineToModify(
```

Comment:
This means that we add our sub-pipeline processors directly to the users pipeline.

If they have a large pipeline and onboard many Unwired streams, this pipeline might become very big, similar for integration pipelines.

I'm wondering if it would be worth here to introduce **another** layer or not.
We could have:
`user/integration pipeline -> @custom pipeline -> @streams pipeline`

and in the `@streams` pipeline we would add references to the sub-pipelines per stream, just like how it works for the unwired streams that don't have a user managed ingest pipeline.

Reply:
Not sure whether another layer of pipelines would help so much with this tbh. Would it simplify something if we did that?

Reply:
No, quite the opposite, it would make it more complicated since there is one more layer to manage and clean up.
I'll ignore it for now, if users complain about their pipelines becoming too polluted then we can deal with it then.